Make QTable Quantity int to float conversion less surprising

orionlee opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

This pull request is to address the issue that during `QTable` `Quantity` conversion, it will convert `int` to `float`.
- add test to make it explicit that the behavior is intentional (rather than a bug)
- Add note to `QTable` apidoc.
- Emit warnings when int-to-float conversion happened. The intent is to catch use cases such as `qtab['col'].unit = u.m`. 

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #10964

orionlee commented:
Regarding the warnings,
- it's emitted in the narrow case that conversion happened during column update, e.g., `qtab['col'].unit = u.m`
- No warning is emitted for the case when a column is added, e.g., adding `Column([1, 2], unit=u.m)`, as it will simply generate many spurious warnings.
- the current implementation (tracking it in an internal state `QTable._dtype_before_conversion`) is a bit hacky. I tried to add the state to the column `qcol` object, its `.info`, and `.meta`, but none of them work (the state was somehow lost).

pllim commented on astropy/table/connect.py:
```diff
@@ -1,5 +1,6 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 
+import warnings
```
Not used?
```suggestion
```

pllim commented on astropy/table/table.py:
```diff
@@ -3593,6 +3593,9 @@ class QTable(Table):
     except that columns with an associated ``unit`` attribute are converted to
     `~astropy.units.Quantity` objects.
 
+    Note: when a column of ``int`` ``dtype`` is converted to ``Quantity``, its
+    ``dtype`` is converted to ``float``.
```
```suggestion
    .. note:: When a column of ``int`` ``dtype`` is converted to ``Quantity``, its
              ``dtype`` is converted to ``float``.
```

pllim commented on astropy/table/table.py:
```diff
@@ -3644,12 +3649,33 @@ def _convert_col_for_table(self, col):
+    def replace_column(self, name, col, copy=True):
+        super().replace_column(name, col, copy)
+        if self._dtype_before_conversion is not None:
+            warnings.warn("dtype is converted to float for Quantity column '{}'. "
+                          "See https://docs.astropy.org/en/stable/api/astropy.table.QTable.html"
+                          "".format(col.info.name))
```
```suggestion
            warnings.warn(f"dtype is converted to float for Quantity column '{col.info.name}'. "
                          "See https://docs.astropy.org/en/stable/api/astropy.table.QTable.html",
                          AstropyUserWarning)
```

mhvk replied:
Remind me: is `AstropyUserWarning` just shown once by default, or every time? I'm not sure we want the every time behaviour here - people will just stop looking at warnings if they happen too frequently...

pllim commented on astropy/table/table.py:
```diff
@@ -3644,12 +3649,33 @@ def _convert_col_for_table(self, col):
+
+    def _set_column_attribute(self, attr, values):
+        # some internal codes use ``_set_column_attribute()`` to assign units to a column,
+        # which might trigger the warnings. Suppress them as ,they are internal codes that
```
```suggestion
        # which might trigger the warnings. Suppress them as they are internal codes that
```

pllim commented on astropy/table/tests/test_column.py:
```diff
@@ -595,7 +595,8 @@ def test_qtable_column_conversion():
     assert isinstance(qtab['f'], table.column.Column)
 
-    qtab['i'].unit = 'km/s'
+    with pytest.warns(UserWarning, match="dtype is converted to float"):
```
```suggestion
    with pytest.warns(AstropyUserWarning, match="dtype is converted to float"):
```

pllim commented on astropy/table/tests/test_table.py:
```diff
@@ -2258,6 +2259,46 @@ def test_table_meta_copy_with_meta_arg():
+
+    tab = QTable(dict(time=[1, 2, 3]))
+    tab['length'] = [9, 8, 7]
+    with pytest.warns(UserWarning, match="dtype is converted to float"):
```
```suggestion
    with pytest.warns(AstropyUserWarning, match="dtype is converted to float"):
```

pllim commented on astropy/table/tests/test_table.py:
```diff
@@ -2258,6 +2259,46 @@ def test_table_meta_copy_with_meta_arg():
+
+    # same for dimensionless unit
+    tab['col2'] = [6, 5, 4]
+    with pytest.warns(UserWarning, match="dtype is converted to float"):
```
```suggestion
    with pytest.warns(AstropyUserWarning, match="dtype is converted to float"):
```

pllim reviewed (commented):
Thanks! I left a general review and a thorough review would be done by subpackage maintainers.

As this is technically a behavior change, it needs a change log under "API change". Thanks!

pllim commented:
p.s. Failures look like they are related to this PR and they need to be fixed. For example:
```
FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
```

orionlee commented:
Fixed wording / formatting, and added changelog entry.

The rest of suggested code changes look reasonable. I'm holding off the changes waiting for more feedback.

Specifically, I myself is not completely sold on the new behavior (emit warning during column update / replacement), and would like to get some feedback.

saimn commented:
ping @mhvk @taldcroft in case you missed this one.

mhvk commented on astropy/table/table.py:
```diff
@@ -3644,12 +3648,33 @@ def _convert_col_for_table(self, col):
             qcol = q_cls(col.data, col.unit, copy=False)
             qcol.info = col.info
             qcol.info.indices = col.info.indices
+            if np.issubdtype(getattr(col, 'dtype', None), np.int):
```
Change `np.int` to `np.integer` (which is the actual type class, while `np.int is int`).

mhvk commented on astropy/table/table.py:
```diff
@@ -3644,12 +3648,33 @@ def _convert_col_for_table(self, col):
             qcol.info = col.info
             qcol.info.indices = col.info.indices
+            if np.issubdtype(getattr(col, 'dtype', None), np.int):
+                self._dtype_before_conversion = col.dtype
```
I think logically the warning should be given here. Is there a particular reason not to do that?

It is risky to use some outside attribute to communicate between methods, so would be good to avoid it!

orionlee replied:
I apologize for the long delay.

Using the attribute is indeed hacky, the part that I am most uneasy about in this PR.

The reason that a warning cannot be issued in here in `_convert_col_for_table()` is that it is used in various other scenarios, in addition to the use case conversion in unit assignment `tab['a'].unit = 'km'` this PR tried to address.

E.g., if a warning is issued here, the following codes will also emit warning, even though from an user's perspective, there is no conversion taken place. `t['b']` is consistently a float-based Quantity. 

```python
t = QTable([[1, 2], [3, 4]], names=['a', 'b'])
t['b'] = Column([4, 5], unit='cm')
```

Any idea on how to address this?

mhvk reviewed (commented):
Definitely can see the benefit of a warning here, but a question about the approach.

pllim commented:
There are conflicts now. Also, we have switched the way we do change log, so you have to undo changes to `CHANGES.rst` and follow instructions at https://github.com/astropy/astropy/blob/main/docs/changes/README.rst

orionlee commented:
@pllim I'd like to gauge some feedback before making more commits that would take care of conflicts, etc, as I am not entirely comfortable with the PR. See https://github.com/astropy/astropy/pull/10992#discussion_r622436271

pllim commented:
In that case, given the impending feature freeze, I'll take the liberty to move milestone. Thanks for your understanding!

orionlee commented:
@pilim One thing we could safely do is to document the existing behavior (without issuing warnings). I could create another PR for that in the next day or so. Let me know if you'd like me to do so.

pllim commented:
> document the existing behavior (without issuing warnings)

@orionlee , yes, that would be nice to have until this PR can progress in the future. Thank you!
